fedimint-api/src/task.rs

```diff
@@ -221,12 +224,35 @@ impl TaskGroup {
         guard.completed = true;
     }
 
-    pub async fn join_all(self) -> anyhow::Result<()> {
-        for (name, join) in self.inner.join.lock().await.drain(..) {
+    pub async fn join_all(self, join_timeout: Option<Duration>) -> anyhow::Result<()> {
+        while let Some((name, join)) = self.inner.join.lock().await.pop_front() {
             debug!("Waiting for {name} task to finish");
-            join.await
-                .map_err(|e| anyhow!("Thread {name} panicked with: {e}"))?;
-            debug!("{name} task finished.");
+
+            #[cfg(not(target_family = "wasm"))]
```

Comment:
I would not bother with timeouts here.

Instead, I would go for:

```
impl TaskGroup {
  fn install_shutdown_deadline_terminator(&self) {
      let s_clone = self.clone();

      task::spawn(|| async { // standalone non-task-group thread
          // wait for shutdown flag being set
          while !s_clone.is_shutting_down() {
                sleep(1).await; // could be waiting for the shutting down callback or something instead of polling, but whatever
          }

          sleep(30).await; // give em some time
          

          if !s_clone.had_successful_shutdown() {
              std::process:exit(-10); // party is over
           }
      ||);
  }
}